### ISSUE TITLE
privacy and self-sovereign identity

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#3013]
_draft thesis direction_

Investigate the revealing of privacy-sensitive attributes of your identity.
Should be usable even in the sensitive medical domain with electronic patient files.

First step, find related work in this area (irma, [zero-knowledge](https://blog.cryptographyengineering.com/2014/11/27/zero-knowledge-proofs-illustrated-primer/), [zkSNARKs](https://blog.ethereum.org/2016/12/05/zksnarks-in-a-nutshell/))

For next meeting: gather all general docs, [weforum](www3.weforum.org/docs/WEF_A_Blueprint_for_Digital_Identity.pdf
) Djuri his thesis, etc. in .PDF form.

**Comment by synctext** [comment]
[Facinating 2012 report from World Economic Forum](http://www3.weforum.org/docs/WEF_IT_RethinkingPersonalData_Report_2012.pdf)

```
To restore trust, this report proposes three separate, but related
questions, which need to be addressed by all stakeholders:

1. Protection and Security: How can personal data be protected
   and secured against intentional and unintentional security
   breach and misuse?
2. Rights and Responsibilities for Using Data: How can rights
   and responsibilities, and therefore appropriate permissions, be
  established for personal data to flow in ways that both respect
  its context and balance the interests of all stakeholders?
3. Accountability and Enforcement: How can organizations
  be held accountable for protecting, securing and using
  personal data, in accordance with the rights and established
  permissions for the trusted flow of data?
```

**Comment by synctext** [comment]
10 principles: http://www.lifewithalacrity.com/2016/04/the-path-to-self-soverereign-identity.html
active group: http://www.weboftrust.info/
paper: https://github.com/WebOfTrustInfo/rebooting-the-web-of-trust-fall2016/blob/master/final-documents/reputation-toolkit.pdf
paper: https://github.com/WebOfTrustInfo/ID2020DesignWorkshop/blob/master/final-documents/physician-patient-relationship.pdf

**Comment by AngelaBarrio** [comment]
**Read literature:**

- The inevitable rise of self-sovereign identity - Sovrin Foundation, 2016
- Towards self-sovereign identity using blockchain technology - D. Baars, 2016
- Portable trust: Biometric-based authentication and blockchain storage for self-sovereign identity systems - Hammudoglu, J. et al., 2017
- Decentralizing privacy: using blockchain to protect personal data - Zyskind et al., 2015
- Healthcare data gateways: foud healthcare intelligence on blockchain with novel privacy risk control - Yue, X. et al., 2016
- A case study for blockchain in healthcare: "medrec" prototype for electronic health records and medical research data - ekblow et al., 2016
- Rethinking personal data: strengthening trust - World economic forum, 2012
- Portable reputation toolkit use cases - Allen et al., 2016
- The path to self-sovereign identity - Allen, C. et al., 2016

**Attended events:**

- Techruption meeting (June 15)
- ECP special Blockchain: wet- en regelgeving (June 29)

**Comment by synctext** [comment]
Please also read detailed (privacy-minded) crypto papers:
- Hawk: The blockchain model of cryptography and privacy-preserving smart contracts
- Efficient Zero-Knowledge Contingent Payments in Cryptocurrencies Without Scripts.
- Rational Zero: Economic Security for Zerocoin with Everlasting Anonymity.
- [Blockchain & infrastructure (identity, data security)](https://www.getsmarter.com/career-advice/wp-content/uploads/2016/12/mit_blockchain_and_infrastructure_report.pdf)

**Comment by synctext** [comment]
[Key paper discussing 15 years of decentralization with 189 citations from the literature {including TribBler}](https://github.com/Tribler/tribler/files/1158367/15_years_decentralization__PETS_2017.pdf)
Titled: "Systematizing Decentralization and Privacy: Lessons from 15 Years of Research and Deployments"

This paper is a beautiful version of 3 years earlier work by TUDelft students https://cryptome.org/2014/04/tor-vulns-attacks.pdf titled "The fifteen year struggle of decentralizing
privacy-enhancing technology"

**Comment by synctext** [comment]
[Real patient open dataset](https://wiki.openmrs.org/display/RES/Demo+Data): 
```A sample anonymized data set, including 5,000 patients and 500,000 observations, is available for the following Platform releases (note that demo data for OpenMRS Platform releases and will not work for OpenMRS Reference Application releases)```

**Comment by synctext** [comment]
@qstokkink can you please provide us with a reference to the Stanford paper which is not that popular but seems to have a the right crypto magic we need for key attestation and zero-knowledge proof fun?

**Comment by qstokkink** [comment]
https://crypto.stanford.edu/%7Edabo/papers/2dnf.pdf

**Comment by AngelaBarrio** [comment]
Read literature:

- Enigma: Decentralized computation platform with guaranteed privacy
- Systematizing Decentralization and Privacy: Lessons from 15 Years of Research and Deployments
- The fifteen year struggle of decentralizing privacy-enhancing technology
- Evaluating 2-DNF Formulas on Ciphertexts

**Comment by AngelaBarrio** [comment]
Read literature:

- Wat moet ik nou met blockchain? - Idius Felix, 2017
- De kwaliteit van het elektronisch patientendossier van huisartsen gemeten - NIVEL,2011
- Several presentations and news articles (MedMij, ZorgLog, UMCG blockchain vision etc)

Meeting with Idius Felix. He advised me to look for another dmain because I lack domain knowledge about healthcare. He thought my research question was too broad.

Tried to contact Ron van den Bosch (UMCG) but he retired this month.

**Comment by AngelaBarrio** [comment]
[research-proposal-seminar.pdf](https://github.com/Tribler/tribler/files/1348684/research-proposal-seminar.pdf)

**Comment by synctext** [comment]
Felix consult; the electronic patient file does not exist. Spread out across hospital, special clinics, general practitioner, elderly care, etc. Future direction:
- broader, beyond medical. Generic attribute hiding.
- narrow, focus on 1 medical case study. Align with the app world.
- narrow, focus on the electronic switching point of EPD.
- narrow, focus on 1 group, like general practitioners. Existing open source: company in this area](http://www.elatro.nl/opensource) and [ideas](http://andra.nl/een-open-source-huisarts-informatie-systeem/).
- Shift to implementation + policy of self-sovereign ID, internal affairs ministry

**Comment by AngelaBarrio** [comment]
Activities:
- Python course at Codecademy (currently at 67%)
- Read 2 papers about TrustChain

**Comment by synctext** [comment]
Medicit input: self-sovereign identity where patient can see all available data, append corrections, control sharing, and control removal. [Defined format and medical patient data](https://www.registratieaandebron.nl/wat-is-registreren-aan-de-bron/de-kern-van-registreren-aan-de-bron/basisgegevensset/). Demo app https://mxi.nl/upload/documenten/presentatie_doorontwikkeling_blockchain_zin_izo_mxi.pdf

**Comment by AngelaBarrio** [comment]
Activities:
- Finished Python course
- Started Android app course
- Read report "Fifth Annual Study on Medical Identity Theft"
- Read paper "Exploring medical identitiy theft"
- Read paper "Cybersecurity in healthcare: A systematic review of modern threats and trends"

**Comment by synctext** [comment]
- _Smartphone becomes your personal medical data server. Problem, reliability, backup, and theft-locking. _
- medical records storage
- no not store on a single device
- decentralised trusted data storage environment: use your friends
- possible solution: threshold encryption. ([1992](https://link.springer.com/chapter/10.1007/3-540-57220-1_47) and "recent" [2006 work](https://link.springer.com/content/pdf/10.1007/11605805.pdf#page=236)

**Comment by qstokkink** [comment]
[Threshold Cryptography C Library](https://github.com/niclabs/tchsm-libtc)
[Practical Threshold Signatures](https://www.iacr.org/archive/eurocrypt2000/1807/18070209-new.pdf)
[A JCA-based Implementation Framework for Threshold Cryptography](https://www.acsac.org/2001/papers/42.pdf)
[Strength in Numbers: Threshold ECDSA
to Protect Keys in the Cloud](https://eprint.iacr.org/2015/1169.pdf)

**Comment by AngelaBarrio** [comment]
Read literature:

- Threshold cryptosystems: Authentication and secret sharing (1992)
- Strength in Numbers: Threshold ECDSA to Protect Keys in the Cloud (2015)

**Comment by synctext** [comment]
Buzzword bingo for possible thesis title and storyline:
- __Blockchain-based distributed tamper-proof filesystem using [threshold encryption](https://eprint.iacr.org/2015/1169.pdf)__
- _Smartphone-based blockchain storage primitive with theft locking, disaster recovery, side-channel protection, and distributed key storage using threshold encryption_
- _Blockchain-based medical file sharing protected using threshold encryption_

Suggested ToDo:
- continue reading papers in more details
- try to find Python library for threshold encryption, try our 2 our of 3 prototype
- try to understand the possibility of going beyond signatures with encrypted block storage

**Comment by synctext** [comment]
- keeping busy with Distributed Algorithms course
- [Student team Shamir secret sharing on Android](https://github.com/Tribler/tribler/issues/3246#issuecomment-351103126)
- elliptic curve detailed reading
- End-goal: still mentioned above, prototype _self-sovereign identity where patient can see all available data, append corrections, control sharing, and control removal._

**Comment by synctext** [comment]
please read https://link.springer.com/chapter/10.1007/978-3-319-67816-0_21
```
4 Definitions and Threat Model
We consider a setting in which users maintain attributes about themselves and require registrars to
vouch for these attributes. For example, in order to register the attribute “over 18 years old,” a user
reveals their identity to the government, who verifies their age.
Definition 1 (Passive/active verification)
Definition 2 (Attribute integrity)
Definition 3 (Attribute privacy)
```

**Comment by synctext** [comment]
Related work on zero-knowledge proof, Nature papers. [Zero-Knowledge Nuclear Warhead Verification](http://nuclearfutures.princeton.edu/projects/zkp/).

**Comment by AngelaBarrio** [comment]
**Activities:**
- Christmas holiday
- Finished Distributed algorithms lab
- Read paper "Who Am I? Secure Identity Registration on Distributed Ledgers" by Azouvi, Al-Bassam and Meiklejohn (2017)
- Started writing 8-page thesis draft. Mainly chapter on related work. Now on 5 pages.

**Comment by AngelaBarrio** [comment]
[thesis_draft.pdf](https://github.com/Tribler/tribler/files/1622962/thesis_draft.pdf)

**Comment by synctext** [comment]
The sole focus of this thesis are the following 2 research questions. Numerous other matters can be directly derived from them.
```
This question can be split into two subquestions:
R1: "How can accountability on access be guaranteed in an EMR?"
R2: "How can entries be validated by a patient as well as a healthcare provider in an EMR?"
```

do you need double-spending protection in your specific application domain, do you thus need to explain it?
Four-part, complex sentence, with a lot of nuance, spread over several parts:
```
All data requests are evaluated and in case of a positive permission,
secure multiparty computation (sMPC) is used to process patient data without
risking patient privacy
```
Thesis chapter options
- Introduction
- The problem of sensitive medical data / Problem Description
- Prior work / background technology / crypto tutorial
- Requirements, specifications, and design choices / Design alternatives
- Detailed design (leaving prior chapter at high-scientific level)
- prototyping and evaluation
- DONE

Questions; where to discuss in thesis: GDPR? no PKI yet; Public key are not used yet after [25 years of trying](https://github.com/Tribler/tribler/issues/2682#issue-195474778), the problem is central servers, Self-Sovereign Identity is not explained yet in Chapter 2-ish, etc.

**Comment by AngelaBarrio** [comment]
Activities:

- Improving thesis draft (processing feedback last time + adding literature references + expanding Related Work chapter)

- Read student papers Blockchain engineering and Tanzania blockchain presentation

Question: how to incorporate SSI in this.
[thesis.pdf](https://github.com/Tribler/tribler/files/1743500/thesis.pdf)

**Comment by synctext** [comment]
- Explore possible use-cases
- get Android apps operational from sources ([overlay, self-sovereign identity, and key recovery](https://github.com/Tribler/tribler/issues?q=is%3Aissue+is%3Aopen+label%3A%22MSc+course+work%22)  
- Thesis core either:
  - tamper-proof records + ZorgBerichtenBox as generic must-have building block
  - prototype for use-case (Medmij, mijnzorglog, actuele medicijn indicatie, etc.)

**Comment by synctext** [comment]
_completed_; in-depth study of related for the topic of:
- tamper-proof access logs
- threshold signatures

Next step with the Trustchain app. [QR codes are being used in the medical domain](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5005258/) ```Emergency personnel access the emergency information by scanning the QR code with a smartphone.```
Commercial provider of [in case of emergency info](https://www.iceid.me/home/consumer_home)
![image](https://user-images.githubusercontent.com/325224/37285513-7ce7050a-25ff-11e8-8ca5-59321218332d.png)

Another project [Health QR code](http://blog.visualead.com/health-qr-code/).
![image](https://user-images.githubusercontent.com/325224/37285238-baccf65a-25fe-11e8-8ea6-8a3a111acc0b.png).

**Comment by synctext** [comment]
[Schengenverklaring of medische verklaring voor medicijnen](https://www.rijksoverheid.nl/onderwerpen/geneesmiddelen/vraag-en-antwoord/medicijnen-mee-op-reis)

**Comment by synctext** [comment]
Direction of thesis: implementation of a stand-alone algorithm (like algorithmics master course) not software development.

**Comment by synctext** [comment]
Clear goals for coming 3 weeks:
- There are numerous examples of hospital leaks out there [9000 records leaked](https://www.fiercehealthcare.com/privacy-security/website-upgrade-leaks-nearly-9-000-records-at-illinois-hospital-as-hacking)
- Getting ready for next step in next sprint of wiring 2-page Problem Description
- currently keep focusing on Python-blockchain coding
  - Explore for coming week the [existing examples](https://github.com/gburiola/snakecoin)
  - Threshold encryption and other key recovery solutions in real code libraries
  - numerous [theoretical design exist for key recovery systems](https://scholar.google.com/scholar?q=key+recovery+system)
  - try towards simple running code

**Comment by AngelaBarrio** [comment]
Activities last 2 weeks:

- started building mini blockchain based on The World's Tiniest Blockchain
- running code though not very spectacular
- now trying to make the blockchain actually distributed
- added small part about Barbie's hospitalization case to the problem description

**Comment by synctext** [comment]
- creating an AngelaChain from scratch
- focus on the distributed aspect, going beyond the starting point of single node example
- getting ready to add threshold encryption or other key recovery
- Found the state-of-the-art technique to improve upon
  - [Strength in Numbers: Threshold ECDSA to Protect Keys in the Cloud](https://eprint.iacr.org/2015/1169.pdf)
  - side-channel protection: private key never appears in memory
  - form of architectural+algorithmic countermeasure
  - does not assume a central trusted entity, still traditional distributed setting
  - key opportunity is to introduce blockchain-magic
  - no code online yet [on their possible github ](https://github.com/vernamlab?tab=repositories) page
- risk: Worcester work contains strong claims, but something was missed...
- performance on 10-core Xeon box seems sufficient, all below 100ms results.

**Comment by AngelaBarrio** [comment]
Activities:

- Read paper "Patient-centered transparency requirements for medical data sharing systems" (Spagnuelo & Lenzini, 2016)
-Expanded thesis introduction, problem statements and related work section
- Tried to 100% understand Green & Eisenbarth TECDSA paper. CONCERN: nothing about signature verification in the paper?
- Attempt to improve my blockchain, not very successful. Made an appointment with a Python/Flask wizard.

**Comment by synctext** [comment]
- 3. trustless operation; no need to trust any government agency or health care technology provider
- connects to https://www.digitaleoverheid.nl/beleid/regie-op-gegevens/
- _validation of database record_, you start to use (signed) irrefutable records. Creation of irrefutable records is a goal-type of formulation. Similar to: https://www.sprout.nl/artikel/crypto/deze-ondernemer-ontwikkelde-een-zorgplatform-op-basis-van-blockchain-met-4-unieke
- ToDo:
  - 1-ish week to gather two or three papers [with distributed secret sharing](https://scholar.google.com/scholar?as_ylo=2014&q=distributed+shamir+secret)
  - read and try "to blockchain" these designs.
  - can it be implemented in Python within scope of a thesis?

**Comment by AngelaBarrio** [comment]
Activities:
- Expanded mini blockchain with Flask functionality that adds the content of a http post to the blockchain
- Read papers "Care. data and access to UK health records: patient privacy and public trust" and "Database State" about the privacy laws compliance of state databases including EMR
- Expanded thesis text in several areas
- Problem: got stuck thinking about software architecture
- Compared 3 EMR system papers, found one useful, one medium and one quite vague
- Useful one is MedRec, uses PyEthereum, maybe switch to that
[thesis_6_juni.pdf](https://github.com/Tribler/tribler/files/2075550/thesis_6_juni.pdf)

**Comment by synctext** [comment]
- MIT media lab student work from 2016: https://ieeexplore.ieee.org/document/7573685/
![image](https://user-images.githubusercontent.com/325224/41030883-9c99237e-697f-11e8-8003-4f720330bb4e.png)
- Healthcare Data Gateway
![image](https://user-images.githubusercontent.com/325224/41030982-da9c849a-697f-11e8-8c4d-49596b45a795.png)

**Comment by maxvisser** [comment]
I subscribed to this post and saw that you maybe start looking at switching to PyEthereum; I see you guys are trying to do these steps:
1. Secret splitting
2. Secret fragment distribution
3. Secret fragment publishing
4. Secret reconstruction

This for me sounds like Shamir Secret Sharing; there are currently two projects in the Ethereum ecosystem working on this; Since 31st of May 2018, Kimono released their smart contract on Rinkeby for trustless secret sharing; kimono medium post: https://medium.com/@pfh/kimono-trustless-secret-sharing-using-time-locks-on-ethereum-8e7e696494d

The kimono platform uses a commit and reveal scheme that distributes the secret for each piece of data across a network of peers who share it with the public only once the time-lock is complete. it relies on [Shamir Secret Sharing](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing)

uPort also has a bounty open for a implementation of shamir secret sharing (sss) in WebAssembly, that is almost complete. https://github.com/uport-project/sss-wasm/pull/2

**Comment by synctext** [comment]
Roadmap for thesis. Concrete tasks
- [x] Thesis direction and Problem Description draft - _Samantha "Barbie" de Jong_
- [x] Related Work - draft
- [x] initial code -  Completely from scratch, without any unneeded features and functionality.
- [ ] Library Joint Random Secret Sharing
- [ ] Chapter 5: Self-sovereign identity prototype
  - related work screenshots
  - offer 2 out of 3 redundancy
- [ ] Chapter 6: Using real medical records with Standard Shamir or "Strength in Numbers: Threshold ECDSA to Protect Keys in the Cloud"
- [ ] FINAL CHAPTER: Select algorithm beyond Standard Sharmir
- [ ] Chapter 7: Implement or improve the algorthm (?by making it dstributed)
- [ ] DONE

**Comment by AngelaBarrio** [comment]
Activities:
- Made a homepage where a user can upload a file. The user can indicate who needs to sign this file. This events gets added to the blockchain.
- The blockchain is displayed on another page
- Made a button to make a ecdsa signature using python-ecdsa library. Still needs to be linked to the actual document.
- Made some additions and improvements to the thesis draft.

**Comment by synctext** [comment]
- 4.3 Digital signature algorithm. _are there also various crypto libraries to choose from_?
- 2.2.1 Accountability on access _are these the requirement for your prototype or research goals_
- Another project: move from approving invoices, towards approving hours and try to create jointly agreed facts. Simplify administration greatly. 
- mention loss recovery or out of thesis scope?
- thesis storyline? 
  - tamper-proof access log
  - Irrefutable, tamper-proof medical records
- make it nice by taking inspiration and color code from [others](https://healthdesign.devpost.com/submissions/11495-medical-record-simple-layout)? Note the [power of anatomy pictures](https://en.wikipedia.org/wiki/Electronic_health_record#/media/File:Electronic_medical_record.jpg)
![image](https://user-images.githubusercontent.com/325224/42382094-855d3cd8-8133-11e8-9f00-14258af96700.png)
![image](https://user-images.githubusercontent.com/325224/42382133-9bb0a236-8133-11e8-9056-23fc99942ce3.png)
![image](https://user-images.githubusercontent.com/325224/42382236-f910717c-8133-11e8-9037-d09ee2a72242.png)

**Comment by AngelaBarrio** [comment]
Activities:
Created basic login page.
Improved prototype by adding working sign functionality. Uploader can check boxes to determine who is required to sign the event. Logged-in user can sign the event, this gets added to the blockchain as well.
Made a start with the access monitoring functionality.
Some small additions to thesis draft.

**Comment by synctext** [comment]
next steps
- making it distributed (e.g. multiple VMs or multiple instances with local sockets)
- define end goal or target features
- performance analysis (correctness is bottleneck, how to make graphs of that?)
- compare with requirements

**Comment by AngelaBarrio** [comment]
Activities last month:
- Created a page where the users can see and download the entries.
- Download-events are added to the blockchain
- Added sign-button (still needs to be connected with existing sign method)
- Made blockchain semi-distributed (there is one master node which hosts the website)
- Changed login page (just finished, was needed for signing functionality).
- Some additions to thesis draft. Tried to find out how existing systems do their logging but they don't respond to my emails.

[thesis_3_sept.pdf](https://github.com/Tribler/tribler/files/2344671/thesis_3_sept.pdf)

**Comment by synctext** [comment]
- "This problem cannot be solved with a smart algorithm: it must be brute forced." add [nuance](https://locklessinc.com/articles/crypto_hash/)
- "This problem cannot be solved with a smart algorithm: it must be brute forced. A
consequence of this is the effectiveness against Sybil attacks. The high costs of creating
pseudonymous identities prevents cheap attacks (Vukoli´c 2015). A drawback
is that it is (obviously) computationally intensive, and therefore uses much energy,
which is an unnecessary strain on the environment" storyline switches from brute-force, to Sybil, and 
back.
- Design choices or System architecture and design space.
- who serves the webpages is an architecture or implementation issue
- Todo: how to increase visibility of the engineering hours of prototyping?
- [others also failed to find a lib](https://crypto.stackexchange.com/questions/54874/ecdsa-threshold-signing). Seems a fancy lib to mention: https://petlib.readthedocs.io/en/latest/
  - https://eprint.iacr.org/2016/013
  - https://eprint.iacr.org/2017/552
- key issue "5.3.1 Using threshold signatures", very hidden.
- Solid progress, so far. Thesis is growing. FINAL CHAPTER... What is the key contribution of thesis? Threshold in a medical (emergency) access scenario? Or creation of efficient threshold EC sigs in the distributed prototype? Performance analysis for 1 million patients?

**Comment by AngelaBarrio** [comment]
Activities last 3 weeks:

- everything works except for retrieving blockchain from another node
- made website ok looking
- changed and added some parts in thesis draft
[thesis-24-sept.pdf](https://github.com/Tribler/tribler/files/2410169/thesis-24-sept.pdf)

**Comment by synctext** [comment]
Performance analysis:
- run in cmdline scripted form
- enhance prototype it does something with x-amount of people in a fixed scenario
- X-axis, increase number of patients, doctors from 1 to 30 (or try 50?)
- Y-Axis, determine CPU usage, memory, number of messages,...

Think about more decentralised approach. Can every medical institute run their own website for login and coordination?

**Comment by AngelaBarrio** [comment]
Activities:

- implemented kind of Proof of Elapsed Time consensus

- coducted some experiments for system without consensus

[thesis-21-nov.pdf](https://github.com/Tribler/tribler/files/2603904/thesis-21-nov.pdf)

**Comment by synctext** [comment]
Progress meeting minutes:
- Accountability on access ***implemented***
- Validation of entries ***implemented***
- no long-lived identities yet for doctor or patient
-  possible inclusion of [blockchain baby](https://www.istandaarden.nl/izo/innovaties/blockchain-mijn-zorg-log/kraamzorg) architecture picture
- draw.io or xfig
- 6. MediTrail prototype _global overview_ / implementation / architecture and implementation
- idea behind "Proof of Elapsed Time consensus" focus of this sprint
  - 1 node is elected to extend the blockchain
  - unique blockchain for each medical patient file
  - random rounds to process pending transaction
  - buffer in node contain the pending transaction, then committed to the blockchain
  - unclear to me what the buffering and commitment-delay offers
- For the purpose of performance evaluation we added a consensus algorithm to our prototype. This consensus mechanism is not designed to add features or business functionality. It is added purely as an enabler for determining the cost of consensus in a concrete use-case such as medical patient file access auditing.

**Comment by AngelaBarrio** [comment]
Activities:

- attached private key to user
- conducted experiments
- improved thesis draft
[thesis-9-dec.pdf](https://github.com/Tribler/tribler/files/2660984/thesis-9-dec.pdf)

**Comment by synctext** [comment]
Meeting minutes:
- 24 pages for first 4 chapters
- Chapter 2 research statement
- 1. Accountability on access: knowing who has accessed the file;
  - 1. The medical record system must provide the patient with accountability mechanisms.
  - 2. The medical record system must provide the patient with evidence regarding permissions history for auditing purposes.
  - 3. The medical record system must provide the patient with evidence of security breaches.
- 2. Validation of EMR entries by both the health care provider and the patient.
- Section 2.4 Requirements:
  - 1. Accountability on access: Every access to an entry in the EMR system is recorded. The log contains information on the name of the user who accessed the file, the name of the file itself, and the timestamp of the event.
  - 2. Validation of entries: A user should be able to sign an entry with a secure digital signature. The digital signatures should be verifiable by anyone in the system.
- Section 5.1 "From requirements to use cases", no link to above research questions
- "The system that is being designed in this master thesis is a prototype" strange defensive wording
- POET: "For a permissioned blockchain, it is a secure and efficient algorithm" mention _unlimited_ identity creation and cheating at the lottery. Mention commitment scheme. Magic Intel silicon can't be the magic solution (e.g. tamper-proof).
- Section "5.2 Blockchains", selecting blockchain technology
- Section "5.4 Digital signature algorithm", encryption implementation
- A block contains it's own hash? "6. Hash of this block"
- FIGURE 6.4: Screenshot of the my logs page (btw never referenced in text)
  - Shows 4 message types without documentation
  - Genesis_block, upload_notification, signature_request, file_access_notification
  - 5th: signature_placement
- Pseudocode is tricky (strict: all variables in italic, defined, and used); for instance, undefined ```SigningBlock```. Magic variable "
- single linked list, so no backwards traversal without (documented) index
- Section 7 lacks intro sentence
- 7.1.1: please integrate these 2 lists
- 7.4 Speed of MediTrail features; performance analysis
- Fig 7.2: not scientific to draw a line through 4 data-points! Plus line smoothing is considered bad practice.
- "Distance to latest signing block" please conduct experiments with larger files; upto 1000.
- 7.5.3 Discussion of file access performance, please integrate into single experiment section

**Comment by AngelaBarrio** [comment]
New thesis draft, processed feedback from last time.
[thesis-7-jan.pdf](https://github.com/Tribler/tribler/files/2733236/thesis-7-jan.pdf)

**Comment by synctext** [comment]
Review notes:
- where is the open source repository?
- thesis is "file oriented" and not 'any digital information' contained within a digital patient database.
- Section 5.3.1, "this high resilience may not be fully needed." perhaps replace with something more technical, like, this solution to the double spending problem in an anonymous network is highly inefficient and not appropriate in our medical context.
- 5.5 Monitoring access to files; please reduce this section or make it more scientific by introducing an "5.5 event-based architecture" where evens can be remotely placed signatures, blockchain block creations, or file access. Void of all expensive continuous polling.
5.3.4 Proof of Elapsed Time; "it is a secure and efficient algorithm" hard claim without citations or mathematical proof! Please explain this simple 'sleep-and-wake-up' algorithm and compare it to leader election in Byzantine Fault Tolerant Systems and Nakamoto consensus.
- "every file that has been uploaded should be downloadable by every user in the system." This formulations sounds very privacy invasive.
- "To create a minimum level of aesthetic appeal" :-)
- 7.4.2 "Unfortunately, performing the same large experiments with the PoET configuration would cost too much time compared to the value of the data."; "no data" entries are unclear. why was this experiment stopped?
- 7.4.3: "Unfortunately, performing the same large experiments with the PoET"; again not a clean experiment.
- The 2 magic parameters determine the PoET performance. Make that more explicit and clear, "which makes sense because the parameters for choosing the sleeping time stay set during the experiments."
- "Using the PoET consensus algorithm with the chosen parameters adds tremendous overhead."; no motivation for magic parameter values. 

Future work ideas&brainstorm: run a test network using [public datasets like a prescribed medication](https://www.kaggle.com/nhs/general-practice-prescribing-data/home) for all doctors in the UK for a whole year. Emulate all hospitals, doctors, and medication prescription events.

**Comment by synctext** [comment]
[On Security Analysis of Proof-of-Elapsed-Time (PoET)](https://link.springer.com/chapter/10.1007/978-3-319-69084-1_19). Solid related work for a security analysis. It introduces a likelyhood of being a cheater.
```The basic idea is to use z-test to check whether a node is generating blocks too fast (winning too frequently in the competition with other nodes for block creation```.

**Comment by synctext** [comment]
No more comments on thesis text and chapter wording. Except usage of Sybil term for cheating in a lottery (7.2.1 Sybil attacks). Advised to make "7.2 Resistance against attacks" a single text paragraph without sub-sub-headings.

Final thesis addition suggestion: link outcome of experimental results with the appropriate theory.

Great additional to theoretical thesis depth:
- PoET consensus results: between 5315 ms; 5864 ms
- can be explained with the theory and z-test
- model parameter; between 1 and 10 second for each blockchain round.

**Comment by AngelaBarrio** [comment]
- I fixed the bug!!! performance is much better now
- Had to re-do almost all experiments due to fixing the bug, but it enabled me to include the experiments that would take too much time previously
- Represented Upload and Read event experiments with box plots
- Implemented some text feedback
[THESIS-10feb.pdf](https://github.com/Tribler/tribler/files/2849304/THESIS-10feb.pdf)

**Comment by synctext** [comment]
- nice addition to depth of thesis
- FIGURE 7.1: Time needed for adding upload blocks (N=100), no consensus (battle with .xls)
- explain standard deviations and size of boxplot.
- mixing seconds and milliseconds
- expand section 7.5

**Comment by AngelaBarrio** [comment]
Presentation draft: 
[presentatie.pptx](https://github.com/Tribler/tribler/files/2873489/presentatie.pptx)

**Comment by synctext** [comment]
[Thesis FINISHED](https://repository.tudelft.nl/islandora/object/uuid:1774b1e9-3860-4a02-adaa-be9a92557169/datastream/OBJ/download) :clap:
